### DISCUSSION TITLE
Capturing of stdout fails

### DISCUSSION THREAD

**Discussion by mrlegohead0x45** [pytest-dev/pytest#9916] [Q&A] [answered]
I am trying to write tests for a function that eventually writes to `sys.stdout` (without explicitly flushing it).

The test I have written is thus:
```py
from cppstream import cout

def test_cout(capsys):
    # with capsys.disabled():
    cout << "Hello World!" # eventually calls sys.stdout.write

    # cout.stream.flush() # explicitly flush
    cap = capsys.readouterr()
    assert cap.out == "Hello World!"
```

However, it is failing unexpectedly, saying that it captured the printing to stdout in the main output, but appears to have failed to make this captured stdout available in the test.

Output on a GitHub Actions machine can be found [here](https://github.com/mrlegohead0x45/cppstream/runs/6278596802?check_suite_focus=true#step:7:7), as well as the source code that I am testing.

Adding the explicit call to sys.stdout.flush() does not change anything locally.

<details>
<summary>Pytest full local output</summary>
<p>

```
$ pytest
=========================================== test session starts ===========================================
platform win32 -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0
rootdir: C:\Users\<me>\programming\python\cppstream
collected 1 item

tests\test_cout.py F                                                                                 [100%]

================================================ FAILURES =================================================
________________________________________________ test_cout ________________________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x000001CCDB905F30>

    def test_cout(capsys):
        # with capsys.disabled():
        cout << "Hello World!"

        # cout.stream.flush()
        # print("Hello World!")
        cap = capsys.readouterr()
>       assert cap.out == "Hello World!"
E       AssertionError: assert '' == 'Hello World!'
E         - Hello World!

tests\test_cout.py:10: AssertionError
------------------------------------------ Captured stdout call -------------------------------------------
Hello World!
========================================= short test summary info =========================================
FAILED tests/test_cout.py::test_cout - AssertionError: assert '' == 'Hello World!'
============================================ 1 failed in 0.22s ============================================
```

</p>
</details>

**Comment by RonnyPfannschmidt** [comment, marked as answer]
while not familiar with the cppstream library, its highly unlikely its using sys.stdout, but rather a lower level, please use fd capture and ensure the c++ streams flush sanely

**Reply by RonnyPfannschmidt** [reply]
after reading the related code, i determined that this is a stream caching issue

the sys.stdout thats stored in the stream is a different object than the sys.stdout the test will see (sys capture requires replacing the attributes on sys)

**Reply by mrlegohead0x45** [reply]
Using `capfd` fails as well, https://github.com/mrlegohead0x45/cppstream/runs/6335570330?check_suite_focus=true
Replacing the `cout._stream` in the test file to `sys.stdout` and using `capsys` works though, although it fails when trying to use a fixture to do this: https://github.com/mrlegohead0x45/cppstream/runs/6335640068?check_suite_focus=true

**Reply by RonnyPfannschmidt** [reply]
i commented in https://github.com/mrlegohead0x45/cppstream/pull/2/files#r867378611

i recommend dynamically fetching the stream for that handler dynamically to account for stdio redirection

**Reply by mrlegohead0x45** [reply]
That works, thanks
